RTL layout fixes for wxMSW and wxGTK3

AliKet opened the pull request:
Currently `wxMSW` is broken in RTL layout:
- Try scrolling `wxGrid` (in griddemo)
- Try switching to Regions screen in drawing sample to see the problem.

I tested this PR in RTL layout and it works correctly for me.

Commit:
wxMSW: Revert commit 288b208a0a4d306afa7c356af6e17105bd24485c

The reverted commit added in 2021-09-28 (Fix setting clipping region
for wxDC with RTL enabled (wxMSW)) because:
- SetDeviceClippingRegion() no longer works correctly (see Regions screen
  in the drawing sample).
- It beaks the wxGrid control rendering (try scrolling the grid window
  horizontally in the griddemo).

Commit:
wxMSW: Remove a hack from wxPaintDC

This hack is no longer needed as GetClippingRect() is working correctly
since a long time. It is added in commit e9f8a82 (in 2007-08-15) to fix
clipping box retrieval (see parent commit 27539df).

Commit:
wxMSW: Fix SetClippingRegion() after reverting commit 288b208a0a

AliKet commented:
> Sorry, I don't have time to look at this right now, but it seems wrong to disable the previously working tests unless we have good reasons to think the tests themselves were incorrect -- do we?

I'll try to look into the failing tests later and try to fix them, but RTL in master is definitely broken! see:
![drawing_rtl_bad](https://github.com/user-attachments/assets/3ddd2e0b-e055-4d45-a1e9-02bab1c3fcce)
![grid_rtl_bad](https://github.com/user-attachments/assets/93321c04-0728-42ef-a1e4-9d9d3b4095e0)

With this PR:
![drawing_rtl_ok](https://github.com/user-attachments/assets/48cff939-930d-4d7a-ae0d-481ac934e116)
![grid_rtl_ok](https://github.com/user-attachments/assets/537d85fd-b249-4442-9192-2ace1cf2c200)

vadz commented:
Yes, it does look broken, but I'm worried about partially but not completely fixing it in 3.3.0, resulting in a different kind of brokenness there and more fixes needed for 3.3.1.

The problem is really that we plan to release it in 2 days and this is a relatively big change and I (don't know about you?) am far from being 100% confident that it's correct...

AliKet commented:
> Yes, it does look broken, but I'm worried about partially but not completely fixing it in 3.3.0, resulting in a different kind of brokenness there and more fixes needed for 3.3.1.
> 
> The problem is really that we plan to release it in 2 days and this is a relatively big change and I (don't know about you?) am far from being 100% confident that it's correct...

It's not urgent for me for this to be included in the upcoming release, and I plan to debug this later to try to get rid of commit [928d7eb](https://github.com/wxWidgets/wxWidgets/pull/25426/commits/928d7ebf2e1c42271b4ac2aca69a7fb08a0e58a4) if possible because it really bothers me. So I suggest adding **work needed**to this as a reminder, TIA!

AliKet commented:
drawing sample with and without commit [508460d](https://github.com/wxWidgets/wxWidgets/pull/25426/commits/508460dc384ea68f27e6a306f0fc53c5192a44e7) and a transformation applied to the DC:

**With:**

![drawing_trans_rtl_bad](https://github.com/user-attachments/assets/d23a8d57-7fd3-4ccf-bd2a-c5da68b51449)

**Without:**

![drawing_trans_rtl_ok_2](https://github.com/user-attachments/assets/e96dbfad-c0ac-4a1f-83da-98600cabe21c)

vadz commented:
I'd really like to merge this but I feel that if we do it now, we'll never reenable the tests later, so let's postpone it a bit until someone has time to understand what is broken here.

aliket83 committed:
wxGTK: Fix SetLayoutDirection() not working with wxMemoryDC

The graphics context must be recreated for SetLayoutDirection() to take effect.

aliket83 committed:
Fix OneDevRegionRTL test case after the changes done to wxMSW and wxGTK3

aliket83 committed:
wxGTK: Fix ScreenToClient() returning incorrect results in RTL layout

aliket83 committed:
wxAUI: Don't mirror hint rectangle in RTL layout

This is unnecessary and incorrect at least under wxMSW and wxGTK3,
because ClientToScreen() already returns the correct rectangle under
these platforms.

AliKet commented:
@vadz @paulcor Is this commit [6e94053](https://github.com/wxWidgets/wxWidgets/pull/25426/commits/6e94053f0477da9be825bcf58d5fa2ec1c34c0b7) looks correct to you? Because it works correctly for me with `wxGTK3` in RTL layout (try moving panes in the auidemo).

**Disclaimer:** I haven't tested it under wxGTK2 because last time I checked the RTL support wasn't great there!

aliket83 committed:
wxGTK: Fix popup menu's position in RTL layout

aliket83 committed:
wxGTK: Fix mouse position when initializing wxMouseEvent in RTL layout

The origin is always in the upper right corner for any type of GtkWidget

aliket83 committed:
Add tests to wxWindow::ClientToScreen() for RTL layout

aliket83 committed:
wxGTK: Fix wxDataViewCtrl rendering in RTL layout

AliKet commented:
@vadz Sorry this PR is no longer `wxMSW` port specific!

Commit:
wxGTK: Fix wxDataViewCustomRenderer in RTL layout

vadz commented on include/wx/gtk/private/event.h:
```diff
@@ -74,11 +74,11 @@ template<typename T> void InitMouseEvent(wxWindowGTK *win,
         // origin in the upper right corner
         GtkAllocation a;
-        gtk_widget_get_allocation(win->m_wxwindow, &a);
+        gtk_widget_get_allocation(win->m_wxwindow ? win->m_wxwindow : win->m_widget, &a);
```
I wonder if we should do this here:
```suggestion
        gtk_widget_get_allocation(win->GetConnectWidget(), &a);
```
?

vadz reviewed (approved):
This looks very nice, I really appreciate getting rid of the hacks in wxMSW and platform checks in the tests, thanks a lot for your work on this!

I'll merge this soon, if you agree that we should use `GetConnectWidget()`, we can/will change this later.
